File: Dynamic Programming/BestTimeToBuyAndSellStockWithTransactionFee.py
"""Your are given an array of integers prices, for which the i-th element is 
the price of a given stock on day i; and a non-negative integer fee representing 
a transaction fee.You may complete as many transactions as you like, but you need 
to pay the transaction fee for each transaction. You may not buy more than 1 share 
of a stock at a time (ie. you must sell the stock share before you buy again.)
Return the maximum profit you can make."""


class Solution(object):
    def maxProfit(self, prices, fee):
        n = len(prices)
        cash, hold = 0, -prices[0]
        for i in range(1,n):
            cash = max(cash,prices[i] + hold - fee)
            hold = max(hold, cash-prices[i])
        return cash

File: Greedy/Best Time to BuyAndSellStockWithTransactionFee.cpp
//Problem Link:- https://leetcode.com/problems/best-time-to-buy-and-sell-stock-with-transaction-fee/

#include<bits/stdc++.h>
using namespace std;

class Solution {
public:
    int maxProfit(vector<int>& nums, int fee) {
        int oss = 0;
        int obs = -nums[0];
        for(int i=1;i<nums.size();i++){
            if (oss - nums[i] > obs){
                obs = oss - nums[i];
            }
            if (obs + nums[i] - fee > oss){
                oss = obs + nums[i] - fee;
            }
        }
        return oss;
    }
};